### PR TITLE
[feat] PL mvp1: validation + logging

### PR DESCRIPTION
**Pull request by ytsheng**
- [x] MVP 1. Evaluation and Logging - Eval accuracy parity between mmf_trainer and lightning_trainer
  -  [X] Validation at the right frequency and at the end (also calculate the right metrics)
  -  [X] Do logging (same logging format with slight difference is ok)
- [x] Tests - tests only done for pytorch lightning integration
  - [X] test logging values are the same as `mmf_trainer`'s logging values
  - [X] test eval values are the same as `mmf_trainer`'s eval values
  - [x] test tensorboard logging the same as `mmf_trainer`'s
- [ ] Blocked:
  - [ ] One Epoch Evaluation Run after train: [PR4945](https://github.com/PyTorchLightning/pytorch-lightning/pull/4945/files) and [PR4948](https://github.com/PyTorchLightning/pytorch-lightning/pull/4948/files)

### CODE REVIEW

File: mmf/trainers/lightning_core/loop_callback.py [hunk 1]
```diff
@@ -32,9 +49,37 @@ def on_train_batch_end(
...
+
     def on_train_end(self, trainer: Trainer, pl_module: LightningModule):
-        # TODO: @sash next PR
-        pass
+        # Only do when run_type has train as it shouldn't happen on validation and
+        # inference runs. Inference will take care of this anyways. Also, don't run
+        # if current iteration is divisble by snapshot interval as it will just
+        # be a repeat
+        if (
+            "train" in self.lightning_trainer.run_type
```
Comment (apsdehal) [inline mmf/trainers/lightning_core/loop_callback.py:70]: At some places we are using `self.lightning_trainer.x` while at some `self.training_config.x`. This would be confusing for an end user inheriting this to understand where is what.

File: mmf/trainers/lightning_core/loop_callback.py [hunk 2]
```diff
@@ -47,3 +92,134 @@ def on_validation_batch_end(
...
+        ), "mmf lightning_trainer supports 1 optimizer per model for now."
+        optimizer = trainer.optimizers[0]
+        return optimizer
+
+    def _save_checkpoint(self, trainer: Trainer):
+        logger.info("Checkpoint time. Saving a checkpoint.")
+        return
+        # TODO: sash Needs implementation - next mvp
+
+    def _get_current_epoch_for_logging(self, trainer: Trainer):
```
Comment (apsdehal) [inline mmf/trainers/lightning_core/loop_callback.py:167]: Shouldn't these be properties?

File: mmf/utils/general.py [hunk 3]
```diff
@@ -333,33 +332,6 @@ def get_sizes_list(dim, chunks):
     return sizes_list
 
 
-def get_max_updates(config_max_updates, config_max_epochs, train_loader, update_freq):
```
Comment (apsdehal) [inline mmf/utils/general.py:336]: As discussed before can we import these here as well and add `# noqa` to it so that flake8 doesn't complain.

File: tests/test_utils.py [hunk 4]
```diff
@@ -136,7 +136,13 @@ def forward(self, prepared_batch):
         batch = prepared_batch[DATA_ITEM_KEY]
         output = self.linear(batch)
         loss = torch.nn.MSELoss()(-1 * output, batch)
-        return {"losses": {"loss": loss}, "logits": output, "input_batch": input_sample}
+        return {
+            "losses": {"loss": loss},
+            "logits": output,
+            "input_batch": input_sample,
+            "dataset_type": "dummpy_dataset_type",
```
Comment (apsdehal) [inline tests/test_utils.py:143]:
```suggestion
            "dataset_type": "dummy_dataset_type",
```

next line as well.

File: mmf/configs/defaults.yaml [hunk 5]
```diff
@@ -147,7 +147,7 @@ trainer:
     # https://pytorch-lightning.readthedocs.io/en/latest/trainer.html#trainer-class-api
     type: lightning
     params:
-        gpus: null
+        gpus: 0
```
[the remark this answers is not in the document]
Comment (ytsheng) [reply mmf/configs/defaults.yaml:150]: This makes it so that the default is idx=0 gpu, as opposed to before, on cpu.

File: mmf/models/base_model.py [hunk 6]
```diff
@@ -205,9 +201,28 @@ def validation_step(self, batch, batch_idx, *args, **kwargs):
...
+            Dict
+        """
+        return self._forward_lightning_step(batch, batch_idx)
+
+    def _forward_lightning_step(self, batch, batch_idx):
         batch = self._ensure_sample_list(batch)
         output = self(batch)
-        # TODO: @sash Implementation coming soon! (next PR)
+        loss_dict = output["losses"]
+        output["loss"] = sum(loss.mean() for loss in loss_dict.values())
```
[the remark this answers is not in the document]
Comment (ytsheng) [reply mmf/models/base_model.py:224]: `loss` is required, `input_batch` is returned here so we can have access to it in the callback function to compute report and metrics.

File: mmf/trainers/core/training_loop.py [hunk 7]
```diff
@@ -10,7 +10,8 @@
 from mmf.common.report import Report
 from mmf.common.sample import to_device
 from mmf.utils.distributed import is_xla
-from mmf.utils.general import clip_gradients, get_max_updates
+from mmf.utils.flow import extract_loss, get_max_updates
```
[the remark this answers is not in the document]
Comment (ytsheng) [reply mmf/trainers/core/training_loop.py:13]: Since now there are two places who are going to use those functions, I think it is much cleaner to move them to a utils file where both can access. I thought calling it flow makes sense because it could be used for both training and evaluation work flow.

File: mmf/trainers/lightning_core/loop_callback.py (hunk 1 above)
Comment (ytsheng) [reply mmf/trainers/lightning_core/loop_callback.py:70]: run_type is a config parameter, so we need to grab the config from the trainer to be able to access this variable. In the init of this class, I added comments to make it clear which is which. I also simplified the logic here by a little bit.

File: mmf/trainers/lightning_core/loop_callback.py (hunk 2 above)
Comment (ytsheng) [reply mmf/trainers/lightning_core/loop_callback.py:167]: I do think properties is the best - since it is just + 1 from the original number.

File: mmf/utils/general.py (hunk 3 above)
Comment (ytsheng) [reply mmf/utils/general.py:336]: ok! Will move them here!

File: tests/test_utils.py (hunk 4 above)
Comment (ytsheng) [reply tests/test_utils.py:143]: oh good catch! 💛

**Review by ytsheng — COMMENTED** [review]
Thanks for review!
